src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp

```diff
@@ -0,0 +1,162 @@
+#ifndef STAN_ANALYZE_MCMC_COMPUTE_POTENTIAL_SCALE_REDUCTION_HPP
+#define STAN_ANALYZE_MCMC_COMPUTE_POTENTIAL_SCALE_REDUCTION_HPP
+
+#include <stan/math/prim/mat.hpp>
+#include <stan/analyze/mcmc/autocovariance.hpp>
+#include <stan/analyze/mcmc/split_chains.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <algorithm>
+#include <cmath>
+#include <vector>
+#include <limits>
+
+namespace stan {
+namespace analyze {
+
+  /**
+   * Computes the potential scale reduction (Rhat) for the specified
+   * parameter across all kept samples.
+   *
+   * See more details in Stan reference manual section "Potential
+   * Scale Reduction". http://mc-stan.org/users/documentation
+   *
+   * Current implementation assumes chains are all of equal size and
+   * draws are stored in contiguous blocks of memory.
+   *
+   * @param draws stores pointers to arrays of chains
+   * @param sizes stores sizes of chains
+   * @return potential scale reduction for the specified parameter
+   */
+  inline
+  double compute_potential_scale_reduction(std::vector<const double*> draws,
+                                           std::vector<size_t> sizes) {
+    int num_chains = sizes.size();
+    size_t num_draws = sizes[0];
+    for ( int chain = 1; chain < num_chains; ++chain ) {
+      num_draws = std::min(num_draws, sizes[chain]);
+    }
+
+    Eigen::VectorXd draw_val = Eigen::VectorXd::Random(num_chains);
+    for ( int chain = 0; chain < num_chains; chain++ ) {
+      Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, 1>>
+        draw(draws[chain], sizes[chain]);
+
+      for ( int n = 0; n < num_draws; n++ ) {
+        if ( !boost::math::isfinite(draw(n)) ) {
+          return std::numeric_limits<double>::quiet_NaN();
+        }
+      }
+
+      if ( draw.isApproxToConstant(draw(0)) ) {
+        draw_val(chain) = draw(0);
+      }
+    }
+
+    if ( draw_val.isApproxToConstant(draw_val(0)) ) {
+      return 1.0;
+    }
+
+    Eigen::VectorXd chain_mean(num_chains);
+    Eigen::VectorXd chain_var(num_chains);
+
+    for ( int chain = 0; chain < num_chains; chain++ ) {
+      Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, 1>>
+        draw(draws[chain], sizes[chain]);
+      chain_mean(chain) = draw.mean();
+      chain_var(chain) = (draw.array() - chain_mean(chain)).square().sum()
```

Comment:
This method of computing the empirical variance is prone to numerical inaccuracy.  We should use a Welford accumulator -- these are implemented in Boost but the code is short enough that it can be done by hand here.  See for example https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/stan/math/prim/mat/fun/welford_var_estimator.hpp.

Reply:
Nice.  

Is it too much for this PR to make this same upgrade in stan-dev/stan's [analyze/mcmc/autocovariance.hpp](https://github.com/stan-dev/stan/blob/f372f98065461b1d7b4b4fe3d1b0bd984138887a/src/stan/analyze/mcmc/autocovariance.hpp#L87) also?

Reply:
Could you make an issue and then put together a PR for updating this code and the code in autocovariance.hpp?  That way we can push this PR as is and contain the better variance estimation to its own PR.

Reply:
Will do.